File: CS 4700 (Programming Languages)/Ruby Assignment/mixin.cpp
#include <iostream>
#include <string>

class MixinBase {
private:
	std::string name;
public:
	virtual void setName(std::string n) { name = n; }
	virtual std::string whoAmI() { return name; }
};

template<class T>
class Employee : public T {
public:
	Employee(std::string name) { T::setName(name); }
	std::string whoAmI() { return T::whoAmI(); }
};

template<class T>
class Manager : public T {
public:
	Manager(std::string name) : T(name){};
	std::string whoAmI() { return T::whoAmI(); }
};

int main()
{
	Employee<MixinBase> emp("Bob");
	Manager<Employee<MixinBase>> mgr("Mary");
	std::cout << emp.whoAmI() << std::endl;
	std::cout << mgr.whoAmI() << std::endl;
}

File: CS 4700 (Programming Languages)/Ruby Assignment/mixin.rb
module Debug
	def who_am_i?
		"#{self.class.name}" +
		"(\##{self.object_id}): #{self.to_s}"
	end
end

class Employee
	include Debug
end
class Manager < Employee
	include Debug
end
emp = Employee.new()
mgr = Manager.new()

puts "\n\nWho are they?"
puts emp.who_am_i?
puts mgr.who_am_i?